packages/react-server/core/components/RootElement.js

```diff
@@ -157,26 +166,36 @@ RootElement.installListener = function(element, listen) {
 }
 
 RootElement.scheduleRender = function(element) {
-	var {listen, when} = (element||{}).props||{};
-	if (!(listen||when)) {
+	var {listen, when, componentLoader, childProps} = (element||{}).props||{};
+	if (!(listen||when||componentLoader||childProps)) {
 		return Q(element).then(RootElement.ensureRootElement);
 	}
 
 	// This is what we'll ultimately resolve our return promise with.
 	// It may be changed by the output of `listen` or `when`.
 	var rendered = element;
+	var componentLoaderDeferred = componentLoader ? componentLoader() : null;
 
 	// Install the listener right away to start gathering props.
 	// It may be a gated emitter, but we want to make sure we squeeze
 	// props out of it from the beginning if it's not.
 	// Finally gate on the `when`.
 	return Q(listen && RootElement.installListener(element, listen))
 		.then(el => el && (rendered = el))
-		.then(() => when)
-		.then(childProps => childProps
-			// merge "when" childProps and "listen" childProps preemptively
-			// to prevent cloneElement shallow merge from clobbering "listen" childProps
-			?React.cloneElement(rendered, {childProps: _.assign({}, rendered.props.childProps || {}, childProps)})
-			:rendered
-		)
+		.then(() => Q.allSettled([when, componentLoaderDeferred]))
+		.then(results => {
+			var [whenResult, loadedComponent] = results;
+			if (whenResult.value || loadedComponent.value || childProps) {
+				// merge in child props from listen, when, and childProps
+				const clonedChildProps = _.assign({}, rendered.props.childProps, whenResult.value, childProps);
+
+				// if we have a component loader specified, copy the resolved component
+				// and render that with the current child as a child of that component
+				const currentChild = rendered.props.children;
+				const childToRender = componentLoaderDeferred ? React.createElement(loadedComponent.value, null, currentChild) : currentChild;
```

Comment:
It looks like this will always create a new element from the `componentLoaderDeferred` if there was one.

Should this only use `loadedComponent.value` on the first pass?

Reply:
Not sure I understand. What do we mean by on the first pass?